### ISSUE TITLE
error when connecting to Mosaiq SQL

### ISSUE THREAD

**Issue by bbarbes** [pymedphys/pymedphys#1621]
This is the first time I use pymedphys.mosaiq.

I'm trying to perform a query in my Mosaiq. When using Microsoft SQL Server 2014 the data I read in the welcome window are 'Authentication: Windows Authentication', 'User name SR03MSQBBDD\Elekta' and 'Password' empty. Server IP address is 10.3.200.153

I tried in python:
```
import pymedphys

connection=pymedphys.mosaiq.connect('10.3.200.153')

```
and I get:
```
Provide a user that only has `db_datareader` access to 'PyMedPhys_SQLLogin_Mosaiq_10.3.200.153:1433/MOSAIQ'
```
I write my username
`SR03MSQBBDD\Elekta`

and I get:
```
Warning: QtConsole does not support password mode, the text you type will be visible.Provide the password for the 'SR03MSQBBDD\Elekta' user on 'PyMedPhys_SQLLogin_Mosaiq_10.3.200.153:1433/MOSAIQ'
```
and I press enter (no pass)

and, finally, I get:
```
Traceback (most recent call last):
  File src\pymssql.pyx:636 in pymssql.connect
  File src\_mssql.pyx:1964 in _mssql.connect
  File src\_mssql.pyx:682 in _mssql.MSSQLConnection.__init__
  File src\_mssql.pyx:1690 in _mssql.maybe_raise_MSSQLDatabaseException
MSSQLDatabaseException: (18452, b'Login failed. The login is from an untrusted domain and cannot be used with Windows authentication.DB-Lib error message 20018, severity 14:\nGeneral SQL Server error: Check messages from the SQL Server\nDB-Lib error message 20002, severity 9:\nAdaptive Server connection failed (10.3.200.153)\nDB-Lib error message 20002, severity 9:\nAdaptive Server connection failed (10.3.200.153)\n')
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File ~\Anaconda3\lib\site-packages\pymedphys\_mosaiq\connect.py:42 in __init__
    self._connection = pymssql.connect(
  File src\pymssql.pyx:642 in pymssql.connect
OperationalError: (18452, b'Login failed. The login is from an untrusted domain and cannot be used with Windows authentication.DB-Lib error message 20018, severity 14:\nGeneral SQL Server error: Check messages from the SQL Server\nDB-Lib error message 20002, severity 9:\nAdaptive Server connection failed (10.3.200.153)\nDB-Lib error message 20002, severity 9:\nAdaptive Server connection failed (10.3.200.153)\n')
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File ~\OneDrive - unav.es\0_programas\schedule\sql.py:12 in <module>
    connection=pymedphys.mosaiq.connect('10.3.200.153')
  File ~\Anaconda3\lib\site-packages\pymedphys\_mosaiq\api.py:88 in connect
    connection = _connect.connect_with_credentials(
  File ~\Anaconda3\lib\site-packages\pymedphys\_mosaiq\connect.py:130 in connect_with_credentials
    connection = Connection(
  File ~\Anaconda3\lib\site-packages\pymedphys\_mosaiq\connect.py:53 in __init__
    raise ValueError(
ValueError: When attempting to connect to MOSAIQ@10.3.200.153:1433 with the SR03MSQBBDD\Elekta user a pymssql.OperationalError was raised
```

Any hint? 

Thanks in advance

**Comment by crcrewso** [comment]
`Login failed. The login is from an untrusted domain`

I think this is the salient error. I'm  not familiar with Mosaiq, as I come from Aria. Is it possible there's a whitelist somewhere limiting connections to only be from known good sources? 

In Aria it's set up so the user has to be in a specific group (local to the machine, and known to both Windows and SQL) and the Domain of the server has to be the same as the user. I wonder if this is also the case for Mosaiq too.

**Comment by bbarbes** [comment]
Thanks a lot. I'll ask to my IT server.

**Comment by SimonBiggs** [comment]
Hi @bbarbes,

If you're still having trouble @dg1an3 might be able to help you with this one. Once you've solved it if you could detail how you fixed it, that'd be amazing.

Having a look at the docs:

https://docs.pymedphys.com/lib/ref/mosaiq.html#connect-and-query-the-mosaiq-database

The username and password are for the Mosaiq SQL database. There is often a "reports user" set up which you might be able to use here.

Cheers,
Simon

**Comment by bbarbes** [comment]
Thanks a lot, @SimonBiggs. At the moment, Mosaiq support does not help us too much... They wants us to pay for a Crystal Report they would kindly do... And they have been five months promising to give us support. So I don't know if there are any "reports user" setup. I will ask, anyway.

**Comment by bbarbes** [comment]
Sorry, I sent the message unfinished.
Thaks for the link. I had already read, but I will do again.
I am waiting for the response of my IT team.
Os course, I will leave my experiences.

**Comment by SimonBiggs** [comment]
> So I don't know if there are any "reports user" setup. I will ask, anyway.

I've found that even the barest of setups of Mosaiq come with this user configured. Also, we were able to ask our site IT to create a new user within the Microsoft SQL database that had the "db_datareader" permissions.

The only reason we used this custom user as opposed to the reports user, was just so that it was clear where the requests were coming from.

**Comment by bbarbes** [comment]
Indeed, when I open Microsoft SQL Server, I read "Authentication: Windows Authentication". I asked Mosaiq for help: I they are able to find a solution, I'll explain it here

**Comment by bbarbes** [comment]
Unfortunately, I haven't received any help from Elekta. They say that they don't recomend any external connection to SQL Mosaiq data base, and they won't provide any help. Besides, my IT service says that the port is opened in the firewall, and they can't help me any more.

**Comment by bbarbes** [comment]
when trying to connect, the message is 
```
Provide a user that only has `db_datareader` access to 'PyMedPhys_SQLLogin_Mosaiq_hostname: 10.3.200.153, port: 1433, database:MOSAIQ, alias: = None, username: imedhospitales.com\xxxx password: xxx:1433/MOSAIQ
```
As I said, my SQL server uses Authentication: Windows Authentication

**Comment by crcrewso** [comment]
I'm going to jump in again and add a couple of thoughts. With our Aria database, there has to be a group set up inside the SQL Server with db_datareader access, and that groups members have to include your AD username. 
An alternate is that the group within the database's membership is a machine local group, say DatabaseReadOnly and on the Windows machine a local group with the same name's members include everyone who needs read access. 

Either solution will allow AD credentials to be authenticated by the SQL server, with appropriate permissions, and you should not need the username or password in the SQL Connection string. (If you're using C# there are nice libraries for this) 

But, both solutions require Elekta set up the read only group from an account that has write and admin access over the MOSAIQ database. If they're not going to support that, then you're dead in the water. Sorry.

**Comment by bbarbes** [comment]
Thank a lot to both of you. I will try again with IT, explaining them the hints you have given to me. I will keep you informed of the result

**Comment by crcrewso** [comment]
> Just to be clear, I've always understood that management of SQL server was something that is handled by each site's IT department, not Elekta service.

I'm not sure about Elekta, but for Varian, our IT staff have administration access to the VM, network and backup of the SQL server; at install the SQL database is locked down to Active Directory accounts that only Varian controls. That limits the adjustment or addition of user rights to only that account. 

With Mosaiq being a [Health Canada](https://health-products.canada.ca/mdall-limh/information.do?companyId_idCompanie=162498&lang=eng) approved piece of software I cannot imagine that Elekta would not do the same.

**Comment by sjswerdloff** [comment]
Sounds like you want to run custom reports.  Perhaps that's the way to express it to Elekta support.
How you generate your reports might be a little unorthodox compared to using Crystal Reports but that could be considered a technical detail... As long as you can define the problem in a way that fits the existing processes you are more likely to get a helpful response.
I think that the standard procedure for enabling custom reports involves construction of a user for the purpose

**Comment by brjdenis** [comment]
Hi, although this is an old thread, I would like to add that I had trouble connecting to Mosaiq database. I was trying to do very simple statistics of approved plans, but I couldn't get pymedphys to work. It would seems that pymssql is not working with Mosaiq in my case, or I don't know how to set connection attributes correctly. After some experimentation I found and alternative with pyodbc. This works:
```
connection = pyodbc.connect(
        'DRIVER={SQL Server};'
        'SERVER=my_ip_address;'
        'DATABASE=MOSAIQ;'
        'UID=my_username;'
        'PWD=my_password;'
        'Trusted_Connection=yes;'
        'readonly=true;'
    )
```
